### ISSUE TITLE
Continuing after ServiceLoader ServiceConfigurationError.

### ISSUE THREAD

**Issue by janbartel** [eclipse/jetty.project#4340]
**jetty-9.4 and beyond**

When using the ServiceLoader, we would like to be able to catch Exceptions from each  call to ServiceLoader.iterator().next() or ServiceLoader.iterator().hasNext() and continue on while there are other services to load.

Currently, most code does something like the following, which will stop iterating when the first exception is thrown:

```java
try
{
  for (SomeService s : ServiceLoader.load(SomeService.class))
  {
    //do stuff
  }
}
catch (Exception e)
{
  //log or throw
}
```

However, according to the ServiceLoader javadoc, it is possible to continue iterating _after_ an exception has occurred:  https://docs.oracle.com/javase/10/docs/api/java/util/ServiceLoader.html#iterator(). It would be desirable to log those Services that cannot be loaded, and continue on with those that are. The problem is how to code that. One ugly attempt could be:

```java
        ServiceLoader<SomeService> loader = ServiceLoader.load(SomeService.class);
        Iterator<SomeService> itor = loader.iterator();
        while (true)
        {
            try
            {
                if (!itor.hasNext())
                    break;
               SomeService s = itor.next();
               //do stuff
            }
            catch (ServiceConfigurationError e)
            {
               //log it
            }
        }
```

**Comment by gregw** [comment]
I think this is screaming out for a ServiceLoaderUtil method.
However I would like to see some kind of protection against looping forever.... but not sure what that can be.... perhaps a loop count of 1000 ?  No other way I can see of testing if the iterator is making progress!
Or perhaps using the spliterator and tryAdvance might be a more robust way?

**Comment by janbartel** [comment]
Maybe we just need a coding pattern? For example the AnnotationConfiguration loads ServletContainerInitializers from the ServiceLoader like this:

``` java
        ServiceLoader<ServletContainerInitializer> loader = ServiceLoader.load(ServletContainerInitializer.class);
        Iterator<ServletContainerInitializer> iter = loader.iterator();
        while (iter.hasNext())
        {
            ServletContainerInitializer sci;
            try
            {
                sci = iter.next();
            }
            catch (Error e)
            {
                // Probably a SCI discovered on the system classpath that is hidden by the context classloader
                LOG.info("Error: {} for {}", e.getMessage(), context);
                LOG.debug(e);
                continue;
            }
           //do stuff
         }
```

**Comment by gregw** [comment]
@janbartel the problem with that pattern is that it doesn't catch the exceptions that are thrown from `iter.hasNext()`, which has caused us problems.  Hence the idea of putting it into a util class where it is done right and can be maintained, updated for all usages.

**Comment by janbartel** [comment]
@gregw ah yes, I remember the problem now.

**Comment by sbordet** [comment]
I am not sure this is a good idea.

The problem is that `ServiceLoader` calls are caller-sensitive.

With JPMS using `ServiceLoaderUtil` may result in the fact that you have to give special visibility (read/open) to `jetty-util` that you don't need when you use `ServiceLoader` directly.

We have had similar issues with `MethodHandle` utility methods, where we need to open up or otherwise use lookups with more permissions than strictly necessary.

I think this should be carefully analyzed before moving it to `jetty-util`.

**Comment by gregw** [comment]
@sbordet this is why the call to ServiceLoader.load is not done within the utility method, but in the caller scope.  The utility method is essentially only wrapping the iterator that is returned, which should be allowed by any caller-sensitive code.

**Comment by sbordet** [comment]
@gregw the iterator is allocated in the call site, but it's a lazy operator. It's when you call `next()` that the service instance is actually created and that is performed in the context of a `jetty-util` class.

However, I checked the JDK code and the caller check is only performed at creation time (`ServiceLoader.load()`) so I _think_ we are good with respect to this.

**Comment by gregw** [comment]
@sbordet the first iteration of this had the load in the util and it did not pass tests.  This passes tests.
We definitely need something like this to avoid all the repeated mistakes (eg catching exceptions from `next()` but not `hasNext()`
Perhaps it could just be a wrapper for the Iterator?   But I don't mind the loadAll() and loadFirst(Predicate) style either as it reads well.

**Comment by sbordet** [comment]
@gregw returning a `Stream` would allow you to access the full `Stream` API, whether you want a `List` or just the first, or filter them, etc. 
The `Stream` could wrap the `Iterator` and take care of the exception handling.

**Comment by joakime** [comment]
I threw together a simple experiment project with ServiceLoader as I was curious as to how the stream API worked.

Project can be found at https://github.com/joakime/serviceloader-experiments

The stream API usage can be found at https://github.com/joakime/serviceloader-experiments/blob/master/drinker/src/main/java/org/eclipse/jetty/demo/Drinker.java

Looking like this ...

``` java
        ServiceLoader.load(BarService.class).stream().forEach((barProvider) ->
        {
            try
            {
                BarService service = barProvider.get();
                if (service == null)
                {
                    System.out.println("No service found.");
                }
                System.out.printf("Service (%s) is type %s%n", service.getClass().getName(), service.getType());
            }
            catch (ServiceConfigurationError error)
            {
                System.out.printf("Service failed to load: %s%n", error.getMessage());
                error.getCause().printStackTrace(System.out);
            }
        });
```

The entire error handling is contained within the call to `barProvider.get()`, no need to worry about `.next()` or `.hasNext()`
I kinda like this stream approach / pattern, and it's already built into the JVM's ServiceLoader class.

**Comment by joakime** [comment]
This works too, to skip errors / nulls during iteration.

``` java
        Optional<BarService> service = ServiceLoader.load(BarService.class).stream()
            .map((provider) ->
            {
                // attempt to load service
                try
                {
                    // will either return a service, throw an error, or return null
                    return provider.get();
                }
                catch (ServiceConfigurationError error)
                {
                    LOG.warn("BarService failed to load", error);
                }
                return null;
            })
            .filter((bar) -> bar != null) // filter out empty / error services
            .findFirst();

        BarService barService = service.orElseThrow(() -> new IllegalStateException("Unable to find a BarService"));
        LOG.info("Using Service ({}) is type [{}]", barService.getClass().getName(), barService.getType());
```

**Comment by lachlan-roberts** [comment]
@joakime I tried using a standard iterator for this test, none of the examples throw in the call to `hasNext()` which is what it says it can do in the javadoc.

> When using the service loader's iterator, the hasNext and next methods will fail with ServiceConfigurationError if an error occurs locating, loading or instantiating a service provider. When processing the service loader's stream then ServiceConfigurationError may be thrown by any method that causes a service provider to be located or loaded.

When we reach the call to `provider.get()` we already have a `ServiceLoader.Provider<T>` so surely we've already done the locating step by then.

**Comment by sbordet** [comment]
@lachlan-roberts I'm thinking that the `Stream` approach is enough. While it may be possible that iterating over the _providers_ may throw, I doubt that's the case.

It's when trying to instantiate the _service_ that an exception may happen, because it's the wrong class version (e.g. ALPN), or some dependency is not available, etc.

I think we should just follow the idiom of using the `Stream` and not the `Iterator`, and if we need to wrap `provider.get()`, then we do that where it is needed.

**Comment by gregw** [comment]
Would a utility method that mapped an Provider<T> to an Optional<T> help with a nice usage of streams, without the need for a verbose try{}catch{}?

**Comment by joakime** [comment]
I went ahead and added more testcases to https://github.com/joakime/serviceloader-experiments that @gregw asked about.

In that example project we have the following jar files, which all have `META-INF/services/org.eclipse.jetty.demo.BarService` files.

* `dive-bar.jar` - valid service at org.eclipse.jetty.demo.DiveBar
* `fancy-bar.jar` - valid service at org.eclipse.jetty.demo.FancyBar
* `lost-their-license-bar.jar` - valid service, but the constructor throws an IllegalStateException
* `dependent-bar.jar` - valid service, but service depends on a class that a missing dependent jar has
* `invalid-bar.jar` - the referenced service class is in the jar, but that class does not implement the service interface making it invalid
* `no-such-bar.jar` - the referenced service class does not exist
* `needy-bar.jar` - the referenced service class exists, but it only has 1 constructor that requires a parameter.

Using the old school iterator method ...

``` java
ServiceLoader serviceLoader = ServiceLoader.load(BarService.class);
Iterator<BarService> serviceIterator = serviceLoader.iterator();
while (serviceIterator.hasNext())
{
    try
    {
        BarService barService = serviceIterator.next();
        if (barService == null)
        {
            LOG.warn("BarService returned null");
        }
        else
        {
            LOG.info("Using Service ({}) is type [{}]", barService.getClass().getName(), barService.getType());
        }
    }
    catch (ServiceConfigurationError error)
    {
        LOG.warn("BarService failed to load", error);
    }
}
```

We see ...

```
[main] WARN org.eclipse.jetty.demo.Drinker - BarService failed to load
java.util.ServiceConfigurationError: org.eclipse.jetty.demo.BarService: Provider org.eclipse.jetty.demo.LostBar could not be instantiated
	at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:581)
	at java.base/java.util.ServiceLoader$ProviderImpl.newInstance(ServiceLoader.java:803)
	at java.base/java.util.ServiceLoader$ProviderImpl.get(ServiceLoader.java:721)
	at java.base/java.util.ServiceLoader$3.next(ServiceLoader.java:1394)
	at org.eclipse.jetty.demo.Drinker.demoOldIteration(Drinker.java:77)
	at org.eclipse.jetty.demo.Drinker.main(Drinker.java:18)
Caused by: java.lang.IllegalStateException: This Bar is Closed (lost our license)
	at org.eclipse.jetty.demo.LostBar.<init>(LostBar.java:7)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.base/java.util.ServiceLoader$ProviderImpl.newInstance(ServiceLoader.java:779)
	... 4 more
[main] INFO org.eclipse.jetty.demo.Drinker - Using Service (org.eclipse.jetty.demo.FancyBar) is type [Fancy]
[main] INFO org.eclipse.jetty.demo.Drinker - Using Service (org.eclipse.jetty.demo.DiveBar) is type [Dive]
[main] WARN org.eclipse.jetty.demo.Drinker - BarService failed to load
java.util.ServiceConfigurationError: org.eclipse.jetty.demo.BarService: org.eclipse.jetty.demo.InvalidBar not a subtype
	at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:588)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1236)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1264)
	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1299)
	at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1384)
	at org.eclipse.jetty.demo.Drinker.demoOldIteration(Drinker.java:73)
	at org.eclipse.jetty.demo.Drinker.main(Drinker.java:18)
[main] WARN org.eclipse.jetty.demo.Drinker - BarService failed to load
java.util.ServiceConfigurationError: org.eclipse.jetty.demo.BarService: org.eclipse.jetty.demo.NeedyBar Unable to get public no-arg constructor
	at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:581)
	at java.base/java.util.ServiceLoader.getConstructor(ServiceLoader.java:672)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1232)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1264)
	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1299)
	at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1384)
	at org.eclipse.jetty.demo.Drinker.demoOldIteration(Drinker.java:73)
	at org.eclipse.jetty.demo.Drinker.main(Drinker.java:18)
Caused by: java.lang.NoSuchMethodException: org.eclipse.jetty.demo.NeedyBar.<init>()
	at java.base/java.lang.Class.getConstructor0(Class.java:3349)
	at java.base/java.lang.Class.getConstructor(Class.java:2151)
	at java.base/java.util.ServiceLoader$1.run(ServiceLoader.java:659)
	at java.base/java.util.ServiceLoader$1.run(ServiceLoader.java:656)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.util.ServiceLoader.getConstructor(ServiceLoader.java:667)
	... 6 more
[main] WARN org.eclipse.jetty.demo.Drinker - BarService failed to load
java.util.ServiceConfigurationError: org.eclipse.jetty.demo.BarService: Provider org.eclipse.jetty.demo.NoSuchBar not found
	at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:588)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.nextProviderClass(ServiceLoader.java:1211)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1220)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1264)
	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1299)
	at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1384)
	at org.eclipse.jetty.demo.Drinker.demoOldIteration(Drinker.java:73)
	at org.eclipse.jetty.demo.Drinker.main(Drinker.java:18)
```

Using the newer `.stream()` methods ...

``` java
ServiceLoader.load(BarService.class).stream().forEach((barProvider) ->
{
    try
    {
        BarService barService = barProvider.get();
        if (barService == null)
        {
            LOG.warn("BarService returned null");
        }
        else
        {
            LOG.info("Using Service ({}) is type [{}]", barService.getClass().getName(), barService.getType());
        }
    }
    catch (ServiceConfigurationError error)
    {
        LOG.warn("BarService failed to load", error);
    }
});
```

We see ...

```
[main] WARN org.eclipse.jetty.demo.Drinker - BarService failed to load
java.util.ServiceConfigurationError: org.eclipse.jetty.demo.BarService: Provider org.eclipse.jetty.demo.LostBar could not be instantiated
	at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:581)
	at java.base/java.util.ServiceLoader$ProviderImpl.newInstance(ServiceLoader.java:803)
	at java.base/java.util.ServiceLoader$ProviderImpl.get(ServiceLoader.java:721)
	at org.eclipse.jetty.demo.Drinker.lambda$demoIterateServices$3(Drinker.java:52)
	at java.base/java.util.ServiceLoader$ProviderSpliterator.tryAdvance(ServiceLoader.java:1491)
	at java.base/java.util.Spliterator.forEachRemaining(Spliterator.java:326)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at org.eclipse.jetty.demo.Drinker.demoIterateServices(Drinker.java:48)
	at org.eclipse.jetty.demo.Drinker.main(Drinker.java:17)
Caused by: java.lang.IllegalStateException: This Bar is Closed (lost our license)
	at org.eclipse.jetty.demo.LostBar.<init>(LostBar.java:7)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.base/java.util.ServiceLoader$ProviderImpl.newInstance(ServiceLoader.java:779)
	... 7 more
[main] INFO org.eclipse.jetty.demo.Drinker - Using Service (org.eclipse.jetty.demo.FancyBar) is type [Fancy]
[main] INFO org.eclipse.jetty.demo.Drinker - Using Service (org.eclipse.jetty.demo.DiveBar) is type [Dive]
Exception in thread "main" java.util.ServiceConfigurationError: org.eclipse.jetty.demo.BarService: org.eclipse.jetty.demo.InvalidBar not a subtype
	at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:588)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1236)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1264)
	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1299)
	at java.base/java.util.ServiceLoader$ProviderSpliterator.tryAdvance(ServiceLoader.java:1483)
	at java.base/java.util.Spliterator.forEachRemaining(Spliterator.java:326)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at org.eclipse.jetty.demo.Drinker.demoIterateServices(Drinker.java:48)
	at org.eclipse.jetty.demo.Drinker.main(Drinker.java:17)
```

Which is interesting, as some of the intentionally bad services don't even appear to be passed to the `.forEach((barProvider) ->`

* `dive-bar.jar` - WORKS normally
* `fancy-bar.jar` - WORKS normally
* `lost-their-license-bar.jar` - EXCEPTION on `provider.get()`
* `dependent-bar.jar` - NOT REPORTED in `forEach()`
* `invalid-bar.jar` - EXCEPTION on `provider.get()`
* `no-such-bar.jar` - NOT REPORTED in `forEach()`
* `needy-bar.jar` - NOT REPORTED in `forEach()`

**Comment by joakime** [comment]
The extra `Optional` isn't necessary, the stream + `ServiceLoader.Provider` does what we need.

**Comment by joakime** [comment]
Also, even with the old school `ServiceLoader` iterator behavior, nothing will throw an exception on `.hasNext()`

**Comment by gregw** [comment]
@joakime I'm sure we have had problems with exceptions thrown from hasNext, that is what started this process in the first place.  Perhaps it is only on some JVMs

**Comment by joakime** [comment]
The `try {} catch (ServiceConfigurationError error) {}` is still going to be usage specific.

Example:

The loading of `ServletContainerInitializers` will log differently if the failure is because access issues between webapp and server classloader.
Which is acceptable and common enough.

And websocket would log at debug/info for Extension load issues, but warn/error (possibly even fatal exception reported) for default Configurator issues.

**Comment by joakime** [comment]
There's only 1 reported issue with ServiceLoader and hasNext().
https://bugs.openjdk.java.net/browse/JDK-8230843

The `hasNext()` issue was due to multiple threads accessing the ServiceLoader (and its iterator) at the same time on early versions of Java 11.
Basically, ServiceLoader is not thread-safe.

If this is a issue for us, using Optional isn't going to fix things, or let us limp along, we have to adjust our usage of the same ServiceLoader (ones for the same service interface) to not access the ServiceLoader from multiple threads.

**Comment by joakime** [comment]
Open question, will ServiceLoader on OSGi with SpiFly still work if we create a utility class?

**Comment by gregw** [comment]
A utility mapping method can make the stream code a little nicer:
```java
    public static <T> Optional<T> provide(ServiceLoader.Provider<T> provider)
    {
        try
        {
            return Optional.of(provider.get());
        }
        catch(Throwable e)
        {
            LOG.warn(e);
            return Optional.empty();
        }
    }

    public void test()
    {
        List<Configuration> configurations = ServiceLoader.load(Configuration.class).stream()
            .map(Configurations::provide)
            .filter(Optional::isPresent)
            .map(Optional::get)
            .collect(Collectors.toList());
    }
```

But the map->filter->map->collect is still a bit verbose.     Perhaps somebody with some better stream-foo than me can come up with a better way to do this, so we can collapse down the try catch but not have too many steps?

**Comment by gregw** [comment]
This is better:
```java

    public static <T> Stream<T> provide(ServiceLoader.Provider<T> provider)
    {
        try
        {
            return Stream.of(provider.get());
        }
        catch(Throwable e)
        {
            LOG.warn(e);
            return Stream.empty();
        }
    }

    public void test()
    {
        List<Configuration> configurations = ServiceLoader.load(Configuration.class).stream()
            .flatMap(Configurations::provide)
            .collect(Collectors.toList());
    }
```

**Comment by joakime** [comment]
No need for `Stream.of()`, or the creation of the `flatMap()`

``` java
    public static <T> T supplies(Supplier<T> supplier)
    {
        try
        {
            return supplier.get();
        }
        catch (Throwable e)
        {
            LOG.warn("Unable to get Service", e);
        }
        return null;
    }

    public void test()
    {
        List<Configuration> configurations = ServiceLoader.load(Configuration.class).stream()
            .map(Configuration::supplies)
            .filter(Objects::nonNull)
            .collect(Collectors.toList());
    }

    public void testClassLoaderSpecific()
    {
        ClassLoader cl = getClassLoader();
        List<Configuration> configurations = ServiceLoader.load(Configuration.class, cl).stream()
            .map(Configuration::supplies)
            .filter(Objects::nonNull)
            .collect(Collectors.toList());
    }
```

**Comment by sbordet** [comment]
@joakime, Greg's version is better, no `null`s around and no need to filter.

**Comment by lachlan-roberts** [comment]
fixed with PR #4602
